core/spec/services/spree/cart/add_item_spec.rb

```diff
@@ -0,0 +1,28 @@
+require 'spec_helper'
+
+module Spree
+  describe Cart::AddItem do
+    subject { described_class }
+
+    let(:order) { create :order }
+    let(:variant) { create :variant, price: 20 }
+    let(:qty) { 1 }
+
+    context 'add line item to order' do
+      it 'change by one and recalculate amount' do
+        expect { subject.call(order: order, variant: variant, quantity: qty) }.to change { order.line_items.count }.by(1)
+        expect(order.amount).to eq 20
+      end
+    end
+
+    context 'with same line item' do
+      let(:line_item) { create :line_item, variant: variant }
+      let(:order) { create :order, line_items: [line_item] }
+
+      it 'not to add' do
+        subject.call(order: order, variant: variant, quantity: qty)
+        expect(order.line_items.count).to eq 1
+      end
+    end
```

Comment:
@gingray let's also test:

* proper passing of options hash - is it recognized and parser properly and how it behaves when an option is wrong
* how it behaves with variants which don't have desired quantity
* how it behaves with non-existing and discontinued variants
* how it behaves when quantity is passed and when quantity isn't passed as param